File: Difficulty/Medium/[153] Find Minimum in Rotated Sorted Array/solution.cpp
/*
 * @lc app=leetcode id=153 lang=cpp
 *
 * [153] Find Minimum in Rotated Sorted Array
 */

// @lc code=start
class Solution {
public:
    int findMin(vector<int>& nums) {
        if (nums.size() < 3) return *min_element(nums.begin(), nums.end());
        int left = 0, right = nums.size() - 1;
        if (nums[left] < nums[right]) return nums[left];
        if (nums[right] < nums[right - 1]) return nums[right];
        while (left <= right) {
            int mid = (left + right) / 2;
            if (nums[mid] < nums[mid + 1] && nums[mid] < nums[mid - 1])
                return nums[mid];
            if (nums[mid] > nums[right])
                left = mid + 1;
            else
                right = mid;
        }
        return 0;
    }
};
// @lc code=end

File: Difficulty/Medium/[153] Find Minimum in Rotated Sorted Array/solution.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        if len(nums) < 3: return min(nums)
        left, right = 0, len(nums) - 1
        if nums[left] < nums[right]:
            return nums[left]
        if nums[right] < nums[right - 1]:
            return nums[right]
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] < nums[mid + 1] and nums[mid] < nums[mid - 1]:
                return nums[mid]
            if nums[mid] > nums[right]:
                left = mid + 1
            else:
                right = mid